controllers/apps/operations/switchover_util.go

```diff
@@ -0,0 +1,463 @@
+/*
+Copyright (C) 2022-2023 ApeCloud Co., Ltd
+
+This file is part of KubeBlocks project
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package operations
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/apecloud/kubeblocks/apis/apps/v1alpha1"
+	componentutil "github.com/apecloud/kubeblocks/controllers/apps/components/util"
+	"github.com/apecloud/kubeblocks/internal/constant"
+	intctrlcomputil "github.com/apecloud/kubeblocks/internal/controller/component"
+	intctrlutil "github.com/apecloud/kubeblocks/internal/controllerutil"
+)
+
+const (
+	SwitchoverCandidateInstanceForAnyPod = "*"
+	SwitchoverCheckJobKey                = "CheckJob"
+	SwitchoverCheckRoleLabelKey          = "CheckRoleLabel"
+)
+
+// needDoSwitchover checks whether we need to perform a switchover.
+func needDoSwitchover(ctx context.Context,
+	cli client.Client,
+	cluster *appsv1alpha1.Cluster,
+	componentSpec *appsv1alpha1.ClusterComponentSpec,
+	switchover *appsv1alpha1.Switchover) (bool, error) {
+	// get the Pod object whose current role label is primary or leader
+	pod, err := getPrimaryOrLeaderPod(ctx, cli, *cluster, componentSpec.Name, componentSpec.ComponentDefRef)
+	if err != nil {
+		return false, err
+	}
+	if pod == nil {
+		return false, nil
+	}
+	switch switchover.InstanceName {
+	case SwitchoverCandidateInstanceForAnyPod:
+		return true, nil
+	default:
+		// If the current instance is already the primary or leader, then no switchover will be performed.
+		if pod.Name == switchover.InstanceName {
+			return false, nil
+		}
```

Comment:
if provided `switchover.InstanceName` is typo it's treated as `SwitchoverCandidateInstanceForAnyPod` which I don't think it's desired behavior, perhaps validate the `switchover.InstanceName` before entering processing path.

Reply:
> if provided `switchover.InstanceName` is typo it's treated as `SwitchoverCandidateInstanceForAnyPod`, which I don't think it's desired behavior, perhaps validate the correctness of `switchover.InstanceName` before entering processing path.

add webhook and controller runtime check done